lib/puppet/pops/merge_strategy.rb

```diff
@@ -0,0 +1,218 @@
+require 'deep_merge/core'
+
+module Puppet::Pops
+  # Merges to objects into one based on an implemented strategy.
+  #
+  class MergeStrategy
+    TypeAsserter = Puppet::Pops::Types::TypeAsserter
+    TypeParser = Puppet::Pops::Types::TypeParser
+
+    # The type used for validation of the _merge_ argument
+    def self.merge_t
+      @@merge_t ||=  TypeParser.new.parse("Variant[String[1],Runtime[ruby,'Symbol'],Hash[Variant[String[1],Runtime[ruby,'Symbol']],Scalar,1]]")
+    end
+    private_class_method :merge_t
+
+    def self.strategies
+      @@strategies ||= {}
+    end
+    private_class_method :strategies
+
+    # Finds the merge strategy for the given key and returns it.
+    #
+    # @param merge_strategy_key [Symbol] The merge strategy key
+    # @return [MergeStrategy] The matching merge strategy
+    #
+    def self.strategy(merge_strategy_key)
+      strategy = strategies[merge_strategy_key]
+      raise ArgumentError, "Unknown merge strategy: '#{merge_strategy_key}'" if strategy.nil?
+      strategy
+    end
+
+    # Returns the list of merge strategy keys known to this class
+    #
+    # @return [Array<Symbol>] List of strategy keys
+    #
+    def self.strategy_keys
+      strategies.keys
+    end
+
+    # Adds a new merge strategy to the map of strategies known to this class
+    #
+    # @param strategy_class [Class<MergeStrategy>] The class of the added strategy
+    #
+    def self.add_strategy(strategy_class)
+      raise ArgumentError, "MergeStrategies.add_strategy 'strategy_class' must be a 'MergeStrategy' class. Got #{strategy_class}" unless MergeStrategy > strategy_class
```

Comment:
The code asks if ```MergeStrategy``` is assignable from the `strategy_class`. It does not ask if `strategy_class` is an instance of ```MergeStrategy```. So I'm a bit reluctant to do that change.